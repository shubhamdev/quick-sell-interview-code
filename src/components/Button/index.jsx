import { bool, func, number, oneOf, oneOfType, string } from "prop-types";

const Button = ({
  label = "",
  width = 60,
  height = "100%",
  color = "primary",
  backgroundColor = "tertiary",
  borderWidth = 1,
  borderStyle = "solid",
  borderColor = "primary",
  onClick = () => {},
  disabled = false,
}) => {
  const colors = {
    primary: "primary",
    secondary: "secondary",
    tertiary: "tertiary",
  };

  const buttonStyles = {
    width,
    height,
    cursor: disabled ? "auto" : "pointer",
    borderWidth,
    borderStyle,
    borderColor: `var(--${colors[borderColor]}-color)`,
    color: `var(--${colors[color]}-color)`,
    backgroundColor: `var(--${colors[backgroundColor]}-color)`,
  };

  return (
    <button style={buttonStyles} onClick={onClick} disabled={disabled}>
      {label}
    </button>
  );
};

Button.propTypes = {
  label: string,
  width: oneOfType([string, number]),
  height: oneOfType([string, number]),
  color: oneOf(["primary", "secondary", "tertiary"]),
  backgroundColor: oneOf(["primary", "secondary", "tertiary"]),
  borderWidth: oneOfType([string, number]),
  borderStyle: string,
  borderColor: oneOf(["primary", "secondary", "tertiary"]),
  onClick: func,
  disabled: bool,
};

export default Button;
