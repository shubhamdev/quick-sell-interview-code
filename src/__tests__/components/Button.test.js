import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import pretty from "pretty";

import Button from "../../components/Button/";

describe("Test Button Component", () => {
  const handleClick = jest.fn();
  it("renders button component", () => {
    const { container } = render(
      <Button
        label="hello"
        width={60}
        height={56}
        borderWidth="0 0 0 1px"
        onClick={handleClick}
      />
    );
    const buttonElement = screen.getByText(/hello/i);
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveStyle({
      width: "60px",
      height: "56px",
      borderWidth: "0 0 0 1px",
      cursor: "pointer",
    });
    userEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(1);

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("renders disabled button component", () => {
    const { container } = render(
      <Button
        label="world"
        width={60}
        height={56}
        borderWidth="0 0 0 1px"
        disabled
        onClick={handleClick}
      />
    );
    const buttonElement = screen.getByText(/world/i);
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toBeDisabled();
    expect(buttonElement).toHaveStyle({
      width: "60px",
      height: "56px",
      borderWidth: "0 0 0 1px",
      cursor: "auto",
    });
    userEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(0);

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });

  it("renders default button component", () => {
    const { container } = render(<Button />);
    const buttonElement = screen.getAllByRole(/button/i)[0];
    expect(buttonElement).toBeInTheDocument();
    expect(buttonElement).toHaveStyle({
      width: "60px",
      height: "100%",
      borderWidth: 1,
      cursor: "pointer",
    });
    userEvent.click(buttonElement);
    expect(handleClick).toHaveBeenCalledTimes(0);

    expect(pretty(container.innerHTML)).toMatchSnapshot();
  });
});
