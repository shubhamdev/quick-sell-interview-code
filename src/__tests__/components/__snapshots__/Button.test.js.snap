// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Button Component renders button component 1`] = `"<button style=\\"width: 60px; height: 56px; cursor: pointer; border-width: 0px 0px 0px 1px; border-style: solid;\\">hello</button>"`;

exports[`Test Button Component renders default button component 1`] = `"<button style=\\"width: 60px; height: 100%; cursor: pointer; border-width: 1px; border-style: solid;\\"></button>"`;

exports[`Test Button Component renders disabled button component 1`] = `"<button style=\\"width: 60px; height: 56px; cursor: auto; border-width: 0px 0px 0px 1px; border-style: solid;\\" disabled=\\"\\">world</button>"`;
